attribution/experiment_logger.py

```diff
@@ -173,14 +176,126 @@ def log_perturbation(
+        self, exp_id: Optional[int] = None, score_agg: Literal["mean", "sum", "last"] = "mean"
+    ):
+        if exp_id == -1:
+            exp_id = self.df_experiments["exp_id"].max()
+
+        token_attrs_df = (
+            self.df_input_token_attribution.groupby(["exp_id", "attribution_strategy"])
+            if exp_id is None
+            else self.df_input_token_attribution[
+                self.df_input_token_attribution["exp_id"] == exp_id
+            ].groupby(["exp_id", "attribution_strategy"])
+        )
+
+        for (exp_id, attr_strat), exp_data in token_attrs_df:
+            if exp_data["input_token_pos"].duplicated().any():
+                exp_data = self._aggregate_attr_score_df(exp_data, score_agg)
+
+            tokens = self.clean_tokens(exp_data["input_token"].tolist())
+            attr_scores = exp_data["attr_score"].tolist()
+            vmax = max(abs(min(attr_scores)), abs(max(attr_scores)))
+            vmin = -vmax
+
+            token_dict = {f"token_{i+1}": t for i, t in enumerate(tokens)}
+            score_dict = {f"token_{i+1}": score for i, score in enumerate(attr_scores)}
+
+            output = self.df_experiments.loc[
+                self.df_experiments["exp_id"] == exp_id, "original_output"
+            ].values[0]
+
+            df = pd.DataFrame([token_dict, score_dict], index=["token", "attr_score"])
+
+            # Generating HTML
+            html_str = '<div style="font-family: monospace;">'
+            for col in df.columns:
+                token = df[col]["token"]
+                score = df[col]["attr_score"]
+                color = self.score_to_color(score, vmin, vmax)
+                html_str += f'<span style="text-decoration: underline; text-decoration-color: {color}; text-decoration-thickness: 4px;">{token}</span>'
+
+            html_str += f" -> <b>{output}</b>"
+            html_str += "</div>"
+
+            # Display
+
+            if get_ipython() and "IPKernelApp" in get_ipython().config:
+                from IPython.display import display
+
+                display(HTML(html_str))
+            else:
+                self.pretty_print(df)
+
+    def print_text_attribution_matrix(self, exp_id: int = -1):
+        if exp_id == -1:
+            exp_id = self.df_experiments["exp_id"].max()
+
+        matrix = self.get_attribution_matrix(exp_id)
+
+        input_tokens = [" ".join(x.split(" ")[:-1]) for x in matrix.index]
+        token_dict = {f"token_{i+1}": t for i, t in enumerate(input_tokens)}
+
+        for oi, output_token in enumerate(matrix.columns):
+            prev_output_str = "".join([" ".join(ot.split(" ")[:-1]) for ot in matrix.columns[:oi]])
+            following_output_str = "".join(
+                [" ".join(ot.split(" ")[:-1]) for ot in matrix.columns[oi + 1 :]]
+            )
+            attr_scores = matrix[output_token].tolist()
+            vmax = max(abs(min(attr_scores)), abs(max(attr_scores)))
+            vmin = -vmax
```

Comment:
Is the idea here to have an upper and lower bound of the range of possible values for the `attr_scores`? If so, do you mind explaining me why we're doing this instead of
```
vmax = max(attr_scores)
vmin = min(attr_scores)
```
?

Reply:
because we want it to be centred around zero (so zero is always a neutral colour, positive attribution is always red and negative is always blue). Otherwise the colour map would be centred around the mean attribution score, which is probably nonzero.